lib/events/api.go

```diff
@@ -579,19 +579,22 @@ type IAuditLog interface {
 	// replay recorded session streams.
 	GetSessionEvents(namespace string, sid session.ID, after int, includePrintEvents bool) ([]EventFields, error)
 
-	// SearchEvents is a flexible way to find events. The format of a query string
-	// depends on the implementing backend. A recommended format is urlencoded
-	// (good enough for Lucene/Solr)
+	// SearchEvents is a flexible way to find events.
 	//
-	// Pagination is also defined via backend-specific query format.
+	// Event types to filter can be specified and pagination is handled by an iterator key that allows
+	// a query to be resumed.
 	//
 	// The only mandatory requirement is a date range (UTC). Results must always
 	// show up sorted by date (newest first)
-	SearchEvents(fromUTC, toUTC time.Time, query string, limit int) ([]EventFields, error)
+	SearchEvents(fromUTC, toUTC time.Time, namespace string, eventTypes []string, limit int, startKey string) ([]EventFields, string, error)
 
-	// SearchSessionEvents returns session related events only. This is used to
-	// find completed session.
-	SearchSessionEvents(fromUTC time.Time, toUTC time.Time, limit int) ([]EventFields, error)
+	// SearchSessionEvents is a flexible way to find session events.
+	// Only session events are returned by this function.
+	// This is used to find completed session.
+	//
+	// Event types to filter can be specified and pagination is handled by an iterator key that allows
+	// a query to be resumed.
+	SearchSessionEvents(fromUTC time.Time, toUTC time.Time, limit int, startKey string) ([]EventFields, string, error)
```

Comment:
These methods should return `[]AuditEvent` instead of `[]EventFields`.  One of the main purposes of moving to GRPC is to get rid of the old dynamic JSON representation wherever we can.

You can use `events.FromOneOf` to convert from an `events.OneOf` to an `events.AuditEvent`.

Reply:
Also I think that this refactor is good, it requires large amounts of internal refactoring which feels out of scope.

Reply:
Perhaps a good Teleport 7.0 target?

Reply:
@xacrimon We're going to need an `UnmarshalAuditEventJSON(...)` helper at some point.  The `api/types/events/oneof.go` file is already full of giant ugly switches.  Thats the nature of the beast in a language without enums I'm afraid.

I think its important that we don't bake deprecated representations into new GRPC APIs.  Not good for the long-term health of the codebase.  I'm fine with continuing to use the old format in tests and the like, but the core logic should be updated to use the new representation.

Reply:
@fspmarshall I started refactoring code to use the typed audit events everywhere and it just seems like too big of a change that I can't say I have time to do properly. There's a ton of testing code to refactor and at least 1-2 kloc of new code that needs to be written to hack around the existing use of the dynamic representation absolutely everywhere along with a refactoring of pretty much all event related code and all event API endpoints that are still on HTTP. Honestly this is not something I can put my word on is right if I rush it. I'd rather perform a gRPC protocol change on a major when we have time refactor this properly. he only way to do this without these other hacky bits is to switch over to gRPC all the way at some point, there's just too much crossreferencing and we would also incur a noticable performance hit converting tens of thousands of events multiple times through the codebase so they fit into the interface API.

Reply:
Okay, actually discovered a much better way to impl this. Updating PR today.

Reply:
Awesome!  Still seeing some `panic("unimplemented")`, so I'll hold off re-reviewing for now.  New changes are looking good from initial pass tho.